File: 백준/Silver/1303. 전쟁 － 전투/전쟁 － 전투.java


import java.util.Arrays;
import java.util.Scanner;

public class Main {

	static int wp; // white power
	static int bp; // blue power
	static char[][] war;
	static boolean[][] visited;
	static int n;
	static int m;

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		m = sc.nextInt();
		n = sc.nextInt();
		war = new char[n][m];
		visited = new boolean[n][m];
		for (int i = 0; i < n; i++) {
			String tmp = sc.next();
			for (int j = 0; j < m; j++) {
				war[i][j] = tmp.charAt(j);
			}
		}
		wp = 0;
		bp = 0;

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (!visited[i][j]) {
					cnt = 1;
					dfs(i, j, war[i][j]);
					if (war[i][j] == 'W') {
						wp += Math.pow(cnt, 2);
					} else {
						bp += Math.pow(cnt, 2);
					}
				}
			}
		}
		System.out.println(wp + " " + bp);

	}

	static int cnt;
	static int[] row = { -1, 1, 0, 0 };
	static int[] col = { 0, 0, 1, -1 };

	private static void dfs(int r, int c, char team) {

		visited[r][c] = true;

		for (int k = 0; k < 4; k++) {
			int nr = r + row[k];
			int nc = c + col[k];
			if (nr >= 0 && nr < n && nc >= 0 && nc < m && !visited[nr][nc] && war[nr][nc] == team) {
				cnt++;
				visited[nr][nc] = true;
				dfs(nr, nc, team);
			}
		}

	}

}

File: 백준/Silver/1303. 전쟁 － 전투/전쟁 － 전투.py
'''
문제 설명
    뭉쳐있으면 제곱만큼 힘을냄
'''
from collections import deque

m,n = map(int, input().split())

grid = [list(input()) for i in range(n)]

w = 0
b = 0

visited = [[False] * m for i in range(n)]

def bfs(r,c,team):
    row = [-1,1,0,0]
    col = [0,0,1,-1]
    q = deque([(r,c)])
    visited[r][c] = True
    cnt = 0
    while q:
        r, c = q.popleft()
        cnt+=1

        for k in range(4):
            nr = r+row[k]
            nc = c+col[k]
            if(not (0<=nr<n and 0<=nc<m)):
                continue
            if not visited[nr][nc] and grid[nr][nc] == team:
                visited[nr][nc] =True
                q.append((nr,nc))
    return cnt

for i in range(n):
    for j in range(m):
        if not visited[i][j]:
            group = bfs(i,j,grid[i][j])
            if(grid[i][j]=="W"):
                w += group**2
            else:
                b+=group**2

print(w,b)